File: Python/0029_Divide_2_Integers.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        is_flip = -1 if (dividend < 0) ^ (divisor < 0) else 1
        dividend, divisor = map(abs, [dividend, divisor])
        
        result = 0
        while dividend >= divisor:
            curr_divisor, count = divisor, 1
            while dividend >= curr_divisor:
                dividend -= curr_divisor
                result += count
                count <<= 1
                curr_divisor <<= 1
        return min(2147483647, result * is_flip)

File: CPlusPlus/0029_Divide_2_Integers.cpp
class Solution {
public:
    int divide(int dividend, int divisor) {
        if (dividend == INT_MIN && divisor == -1) {
            return INT_MAX;
        }
        int needToFlip = (dividend < 0) ^ (divisor < 0) ? -1 : 1;
        long dvd = labs(dividend);
        long div = labs(divisor);
        long result = 0;

        while (dvd >= div) {
            long currDivisor = div;
            long count = 1;
            while (dvd >= currDivisor) {
                result += count;
                dvd -= currDivisor;
                count <<= 1;
                currDivisor <<= 1;
            }
        }
        return result * needToFlip;
    }
};
